# sc-RNAseq-colon analysis

Project for Transcriptomics Course 2021 (MSc Bioinformatics for Computational Genomics) helded by Prof. Giulio Pavesi at Università degli Studi di Milano.

The vignette can be viewed here [GitHub page](https://mariachiaragrieco.github.io/sc-RNAseq-colon/)



## **Description**
The aim of this project is to perform a single cell RNA-Seq study for finding and characterizing cell subtypes

Single-cell RNA-Seq data are retrieved from [PanglaoDB](https://panglaodb.se/index.html) analyzed in this study [“Structural Remodeling of the Human Colonic Mesenchyme in Inflammatory Bowel Disease” (Kinchen J. et al. 2018)](https://pubmed.ncbi.nlm.nih.gov/30270042/) aimed to define how the colonic mesenchyme remodels to fuel inflammation and barrier dysfunction in IBD (Inflammatory Bowel Disease).

The general workflow of the computational analysis here described is based on [Seurat](https://satijalab.org/seurat/) and follows the [Seurat vignette](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html), to find “well defined” clusters to which assign a cell type based on the DE/marker genes found in each.

## References

"Kinchen J, Chen HH, Parikh K, et al. Structural Remodeling of the Human Colonic Mesenchyme in Inflammatory Bowel Disease." Cell. 2018;175(2):372-386.e17. doi:10.1016/j.cell.2018.08.067

"ScRNA-Seq Analysis Workflow." (Hillje, R. 2020.) GitHub Repository. https://romanhaa.github.io/projects/scrnaseq_workflow/; GitHub.



